File: KuantumKaosYonetimi.js
const readline = require('readline'); // Konsoldan veri okumak için kütüphane

// =============================================================
// 1. BÖLÜM: ÖZEL HATA YÖNETİMİ
// =============================================================
// JS'de 'Error' sınıfından miras alıyoruz.
//
class KuantumCokusuError extends Error {
    constructor(nesneID) {
        super(`DİKKAT! ${nesneID} kimlikli nesne çöktü! Patlama gerçekleşti.`);
        this.name = "KuantumCokusuError";
    }
}

// =============================================================
// 2. BÖLÜM: TEMEL YAPI (ABSTRACT CLASS SİMÜLASYONU)
// =============================================================
// JS'de 'abstract' kelimesi yoktur. Bunu 'new.target' ile kontrol ederiz.
//
class KuantumNesnesi {
    constructor(id) {
        // Eğer birisi "new KuantumNesnesi()" derse hata veriyoruz.
        if (new.target === KuantumNesnesi) {
            throw new Error("KuantumNesnesi soyut bir sınıftır, doğrudan üretilemez!");
        }
        this.id = id;
        this.tehlikeSeviyesi = 0;
        this._stabilite = 100.0; // Gizli değişken (Convention olarak _ kullanılır)
    }

    // KAPSÜLLEME (ENCAPSULATION)
    //
    get stabilite() {
        return this._stabilite;
    }

    set stabilite(value) {
        if (value > 100) {
            this._stabilite = 100;
        } else if (value <= 0) {
            this._stabilite = 0;
            // Hata fırlatma
            //
            throw new KuantumCokusuError(this.id);
        } else {
            this._stabilite = value;
        }
    }

    // Soyut Metot Simülasyonu
    // Alt sınıflar bunu ezmezse (override) hata veririz.
    //
    analizEt() {
        throw new Error("analizEt() metodu alt sınıf tarafından uygulanmalıdır!");
    }

    // Durum Bilgisi
    //
    durumBilgisi() {
        // toFixed(2) virgülden sonra 2 basamak gösterir
        return `[${this.id}] Stabilite: %${this.stabilite.toFixed(2)} | Tehlike: ${this.tehlikeSeviyesi}`;
    }
}

// =============================================================
// 3. BÖLÜM: NESNE ÇEŞİTLERİ
// =============================================================

// A. Veri Paketi
//
class VeriPaketi extends KuantumNesnesi {
    constructor(id) {
        super(id);
        this.tehlikeSeviyesi = 1;
    }

    analizEt() {
        this.stabilite -= 5;
        console.log(`${this.id} içeriği okundu. (Stabilite -5)`);
    }
}

// B. Karanlık Madde
class KaranlikMadde extends KuantumNesnesi {
    constructor(id) {
        super(id);
        this.tehlikeSeviyesi = 5;
    }

    analizEt() {
        this.stabilite -= 15;
        console.log(`${this.id} analiz edildi. Karanlık enerji yayılıyor! (Stabilite -15)`);
    }

    // IKritik Arayüzü Metodu (JS'de interface olmadığı için direkt yazıyoruz)
    acilDurumSogutmasi() {
        this.stabilite += 50;
        console.log(`${this.id} soğutuldu. Stabilite yenilendi.`);
    }
}

// C. Anti Madde
class AntiMadde extends KuantumNesnesi {
    constructor(id) {
        super(id);
        this.tehlikeSeviyesi = 10;
    }

    analizEt() {
        console.log("UYARI: Evrenin dokusu titriyor...");
        this.stabilite -= 25;
        console.log(`${this.id} analiz edildi. (Stabilite -25)`);
    }

    // IKritik Arayüzü Metodu
    acilDurumSogutmasi() {
        this.stabilite += 50;
        console.log(`${this.id} soğutuldu. Kritik seviye düşürüldü.`);
    }
}

// =============================================================
// 4. BÖLÜM: OYNANIŞ DÖNGÜSÜ (MAIN LOOP)
// =============================================================

// Konsol okuma arayüzü
const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

const envanter = []; // Listemiz
let sayac = 1;

console.log("--- OMEGA SEKTÖRÜ KUANTUM VERİ AMBARI (JS) ---");

// Menüyü tekrar tekrar göstermek için fonksiyon
function menuGoster() {
    console.log("\n=== KUANTUM AMBARI KONTROL PANELİ ===");
    console.log("1. Yeni Nesne Ekle");
    console.log("2. Tüm Envanteri Listele");
    console.log("3. Nesneyi Analiz Et");
    console.log("4. Acil Durum Soğutması Yap");
    console.log("5. Çıkış");
    
    rl.question("Seçiminiz: ", (secim) => {
        try {
            islemYap(secim);
        } catch (error) {
            //
            if (error instanceof KuantumCokusuError) {
                console.log("\n**************************************");
                console.log("SİSTEM ÇÖKTÜ! TAHLİYE BAŞLATILIYOR...");
                console.log(`Sebep: ${error.message}`);
                console.log("**************************************");
                rl.close(); // Programı kapat
                process.exit(1);
            } else {
                console.log(`Beklenmedik hata: ${error.message}`);
                menuGoster(); // Menüye dön
            }
        }
    });
}

function islemYap(secim) {
    if (secim === "1") {
        const zar = Math.floor(Math.random() * 3) + 1; // 1-3 arası sayı
        const yeniId = `NESNE-${sayac++}`;

        if (zar === 1) envanter.push(new VeriPaketi(yeniId));
        else if (zar === 2) envanter.push(new KaranlikMadde(yeniId));
        else envanter.push(new AntiMadde(yeniId));

        console.log(`${yeniId} ambara kabul edildi.`);
        menuGoster();
    } 
    else if (secim === "2") {
        console.log("\n--- ENVANTER DURUMU ---");
        if (envanter.length === 0) console.log("Ambar boş.");
        
        envanter.forEach(nesne => {
            console.log(nesne.durumBilgisi());
        });
        menuGoster();
    } 
    else if (secim === "3") {
        rl.question("Analiz edilecek ID: ", (girilenId) => {
            try {
                const hedef = envanter.find(x => x.id === girilenId);
                if (hedef) {
                    hedef.analizEt();
                } else {
                    console.log("Nesne bulunamadı!");
                }
            } catch (error) {
                // Analiz sırasında patlama olursa hatayı yukarı fırlat
                if (error instanceof KuantumCokusuError) throw error; 
            }
            menuGoster();
        });
    } 
    else if (secim === "4") {
        rl.question("Soğutulacak ID: ", (girilenId) => {
            const hedef = envanter.find(x => x.id === girilenId);

            if (hedef) {
                // Type Checking (Duck Typing)
                // "Eğer acilDurumSogutmasi diye bir fonksiyonu varsa, bu kritiktir."
                //
                if (typeof hedef.acilDurumSogutmasi === 'function') {
                    hedef.acilDurumSogutmasi();
                } else {
                    console.log("Bu nesne soğutulamaz! (IKritik değil)");
                }
            } else {
                console.log("Nesne bulunamadı!");
            }
            menuGoster();
        });
    } 
    else if (secim === "5") {
        console.log("Çıkış yapılıyor...");
        rl.close();
    } 
    else {
        console.log("Geçersiz seçim.");
        menuGoster();
    }
}

// Başlat
menuGoster();

File: KuantumKaosYonetimi.java
import java.util.ArrayList;
import java.util.List;
import java.util.Random;
import java.util.Scanner;

// 1. BÖLÜM: ÖZEL HATA YÖNETİMİ
class KuantumCokusuException extends RuntimeException {
    public KuantumCokusuException(String nesneID) {
        super("DİKKAT! " + nesneID + " kimlikli nesne çöktü! Patlama gerçekleşti.");
    }
}

// 2. BÖLÜM: ARAYÜZ (INTERFACE)
interface IKritik {
    void acilDurumSogutmasi();
}

// 3. BÖLÜM: TEMEL YAPI (ABSTRACT CLASS)
abstract class KuantumNesnesi {
    protected String id;
    protected int tehlikeSeviyesi;
    private double stabilite;

    // Kurucu Metot (Constructor)
    public KuantumNesnesi(String id) {
        this.id = id;
    }

    public String getId() {
        return id;
    }

    public int getTehlikeSeviyesi() {
        return tehlikeSeviyesi;
    }

    // Encapsulation (Kapsülleme)
    public double getStabilite() {
        return stabilite;
    }

    public void setStabilite(double value) {
        if (value > 100) {
            this.stabilite = 100;
        } else if (value <= 0) {
            this.stabilite = 0;
            // Stabilite 0 veya altına düşerse hata fırlat
            throw new KuantumCokusuException(this.id);
        } else {
            this.stabilite = value;
        }
    }

    // Abstract Metot
    public abstract void analizEt();

    // Durum Bilgisi Metodu 
    public String durumBilgisi() {
        // String.format ile virgülden sonra 2 basamak gösteriyoruz
        return String.format("[%s] Stabilite: %% %.2f | Tehlike: %d", id, stabilite, tehlikeSeviyesi);
    }
}

// 4. BÖLÜM: NESNE ÇEŞİTLERİ (INHERITANCE) 

// A. Veri Paketi 
class VeriPaketi extends KuantumNesnesi {
    public VeriPaketi(String id) {
        super(id);
        setStabilite(100);
        this.tehlikeSeviyesi = 1;
    }

    @Override
    public void analizEt() {
        setStabilite(getStabilite() - 5);
        System.out.println(getId() + " içeriği okundu. (Stabilite -5)");
    }
}

// B. Karanlık Madde 
class KaranlikMadde extends KuantumNesnesi implements IKritik {
    public KaranlikMadde(String id) {
        super(id);
        setStabilite(100);
        this.tehlikeSeviyesi = 5;
    }

    @Override
    public void analizEt() {
        setStabilite(getStabilite() - 15);
        System.out.println(getId() + " analiz edildi. Karanlık enerji yayılıyor! (Stabilite -15)");
    }

    @Override
    public void acilDurumSogutmasi() {
        setStabilite(getStabilite() + 50);
        System.out.println(getId() + " soğutuldu. Stabilite yenilendi.");
    }
}

// C. Anti Madde 
class AntiMadde extends KuantumNesnesi implements IKritik {
    public AntiMadde(String id) {
        super(id);
        setStabilite(100);
        this.tehlikeSeviyesi = 10;
    }

    @Override
    public void analizEt() {
        System.out.println("UYARI: Evrenin dokusu titriyor...");
        setStabilite(getStabilite() - 25);
        System.out.println(getId() + " analiz edildi. (Stabilite -25)");
    }

    @Override
    public void acilDurumSogutmasi() {
        setStabilite(getStabilite() + 50);
        System.out.println(getId() + " soğutuldu. Kritik seviye düşürüldü.");
    }
}

// 5. BÖLÜM: OYNANIŞ DÖNGÜSÜ (MAIN LOOP) 
public class KuantumKaosYonetimi {
    public static void main(String[] args) {
        List<KuantumNesnesi> envanter = new ArrayList<>(); // Generic List 
        Scanner scanner = new Scanner(System.in);
        Random rnd = new Random();
        int sayac = 1;

        System.out.println("--- OMEGA SEKTÖRÜ KUANTUM VERİ AMBARI (JAVA) ---");

        try {
            while (true) { // Sonsuz Döngü 
                System.out.println("\n=== KUANTUM AMBARI KONTROL PANELİ ===");
                System.out.println("1. Yeni Nesne Ekle");
                System.out.println("2. Tüm Envanteri Listele");
                System.out.println("3. Nesneyi Analiz Et");
                System.out.println("4. Acil Durum Soğutması Yap");
                System.out.println("5. Çıkış");
                System.out.print("Seçiminiz: ");

                String secim = scanner.nextLine();

                if (secim.equals("1")) {
                    int zar = rnd.nextInt(3) + 1; // 1, 2 veya 3
                    String yeniId = "NESNE-" + sayac++;

                    if (zar == 1) envanter.add(new VeriPaketi(yeniId));
                    else if (zar == 2) envanter.add(new KaranlikMadde(yeniId));
                    else envanter.add(new AntiMadde(yeniId));

                    System.out.println(yeniId + " ambara kabul edildi.");
                } 
                else if (secim.equals("2")) {
                    System.out.println("\n--- ENVANTER DURUMU ---");
                    if (envanter.isEmpty()) System.out.println("Ambar boş.");
                    
                    for (KuantumNesnesi nesne : envanter) {
                        System.out.println(nesne.durumBilgisi()); // Polimorfizm 
                    }
                } 
                else if (secim.equals("3")) {
                    System.out.print("Analiz edilecek ID: ");
                    String girilenId = scanner.nextLine();
                    
                    // Nesneyi bulma
                    KuantumNesnesi hedef = null;
                    for (KuantumNesnesi k : envanter) {
                        if (k.getId().equals(girilenId)) {
                            hedef = k;
                            break;
                        }
                    }

                    if (hedef != null) {
                        hedef.analizEt();
                    } else {
                        System.out.println("Nesne bulunamadı!");
                    }
                } 
                else if (secim.equals("4")) {
                    System.out.print("Soğutulacak ID: ");
                    String girilenId = scanner.nextLine();
                    
                    KuantumNesnesi hedef = null;
                    for (KuantumNesnesi k : envanter) {
                        if (k.getId().equals(girilenId)) {
                            hedef = k;
                            break;
                        }
                    }

                    if (hedef != null) {
                        // Type Checking (instanceof) 
                        if (hedef instanceof IKritik) {
                            ((IKritik) hedef).acilDurumSogutmasi();
                        } else {
                            System.out.println("Bu nesne soğutulamaz! (IKritik değil)");
                        }
                    } else {
                        System.out.println("Nesne bulunamadı!");
                    }
                } 
                else if (secim.equals("5")) {
                    System.out.println("Çıkış yapılıyor...");
                    break;
                }
            }
        } catch (KuantumCokusuException e) {
            // Game Over Yakalama 
            System.out.println("\n**************************************");
            System.out.println("SİSTEM ÇÖKTÜ! TAHLİYE BAŞLATILIYOR...");
            System.out.println("Sebep: " + e.getMessage());
            System.out.println("**************************************");
        } catch (Exception e) {
            System.out.println("Beklenmedik hata: " + e.getMessage());
        }
        
        scanner.close();
    }
}